flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/MergeTreeWriter.java

```diff
@@ -122,48 +139,61 @@ public void write(KeyValue kv) throws Exception {
 
     @Override
     public long memoryOccupancy() {
-        return memTable.memoryOccupancy();
+        return writeBuffer.memoryOccupancy();
     }
 
     @Override
     public void flushMemory() throws Exception {
-        if (memTable.size() > 0) {
+        boolean success = writeBuffer.flushMemory();
+        if (!success) {
+            flushWriteBuffer();
+        }
+    }
+
+    private void flushWriteBuffer() throws Exception {
+        if (writeBuffer.size() > 0) {
             if (compactManager.shouldWaitCompaction()) {
                 // stop writing, wait for compaction finished
                 trySyncLatestCompaction(true);
             }
 
-            // write changelog file
-            //
-            // NOTE: We must first call memTable.rawIterator(), then call memTable.mergeIterator().
-            // Otherwise memTable.rawIterator() will generate sorted, but not yet merged, data. See
-            // comments on MemTable for more details.
-            if (changelogProducer == ChangelogProducer.INPUT) {
-                KeyValueDataFileWriter changelogWriter = writerFactory.createChangelogFileWriter(0);
-                changelogWriter.write(memTable.rawIterator());
-                changelogWriter.close();
-                changelogFiles.add(changelogWriter.result());
+            final RollingFileWriter<KeyValue, DataFileMeta> changelogWriter =
+                    changelogProducer == ChangelogProducer.INPUT
+                            ? writerFactory.createRollingChangelogFileWriter(0)
+                            : null;
```

Comment:
Create `changelogWriter` when creating `MergeTreeWriter`. Then in `MergeTreeWriter#write` we write into `changelogWriter` and `writeBuffer` at the same time. In this way we can keep the order of changelog.

Reply:
We can not keep the order of changelog now, the min max key will be wrong.

Reply:
Then add comments to `KeyValueDataFileWriter`, stating that records given to the writer must be sorted because we're not comparing the min max keys.